File: backend/src/pipeline/cleanup.js
import { config } from '../config.js';
import { deletePdf } from './pdf.js';

/**
 * runCleanup(prisma) -> { cleanedRecords, cleanedFiles }
 *
 * Finds closed tenders whose endDate is more than
 * config.autoDeleteClosedAfterDays old, deletes their PDF files, and
 * (if archive mode is on) writes a lightweight ArchivedTender record
 * before deleting the Tender document.
 *
 * Never touches tenders where status !== "closed", regardless of age.
 */
export async function runCleanup(prisma) {
  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - config.autoDeleteClosedAfterDays);

  const stale = await prisma.tender.findMany({
    where: {
      status: 'closed',
      endDate: { lt: cutoff },
    },
  });

  let cleanedRecords = 0;
  let cleanedFiles = 0;

  for (const tender of stale) {
    if (tender.pdfPath) {
      deletePdf(tender.pdfPath);
      cleanedFiles += 1;
    }

    if (config.archiveMode) {
      await prisma.archivedTender.create({
        data: {
          source: tender.source,
          bidNumber: tender.bidNumber,
          title: tender.title,
          locationCity: tender.locationCity,
          endDate: tender.endDate,
          bidValue: tender.bidValue,
          emdAmount: tender.emdAmount,
          archivedAt: new Date(),
        },
      });
    }

    await prisma.tender.delete({ where: { id: tender.id } });
    cleanedRecords += 1;
  }

  return { cleanedRecords, cleanedFiles };
}

File: backend/app/pipeline/cleanup.py
"""
app/pipeline/cleanup.py

Direct port of src/pipeline/cleanup.js.
"""
from datetime import datetime, timezone, timedelta

from ..config import config
from .pdf import delete_pdf


async def run_cleanup(tenders_col, archived_tenders_col) -> dict:
    """runCleanup(prisma) -> { cleanedRecords, cleanedFiles }

    Finds closed tenders whose endDate is more than
    config.auto_delete_closed_after_days old, deletes their PDF files, and
    (if archive mode is on) writes a lightweight ArchivedTender record
    before deleting the Tender document.

    Never touches tenders where status != "closed", regardless of age.
    """
    cutoff = datetime.now(timezone.utc) - timedelta(days=config.auto_delete_closed_after_days)

    stale_cursor = tenders_col.find({'status': 'closed', 'endDate': {'$lt': cutoff}})
    stale = await stale_cursor.to_list(length=None)

    cleaned_records = 0
    cleaned_files = 0

    for tender in stale:
        if tender.get('pdfPath'):
            others = await tenders_col.count_documents({
                'pdfPath': tender['pdfPath'],
                '_id': {'$ne': tender['_id']},
            })
            if others == 0:
                delete_pdf(tender['pdfPath'])
                cleaned_files += 1

        if config.archive_mode:
            await archived_tenders_col.insert_one({
                'source': tender.get('source'),
                'bidNumber': tender.get('bidNumber'),
                'title': tender.get('title'),
                'locationCity': tender.get('locationCity'),
                'endDate': tender.get('endDate'),
                'bidValue': tender.get('bidValue'),
                'emdAmount': tender.get('emdAmount'),
                'archivedAt': datetime.now(timezone.utc),
            })

        await tenders_col.delete_one({'_id': tender['_id']})
        cleaned_records += 1

    return {'cleanedRecords': cleaned_records, 'cleanedFiles': cleaned_files}
